### DISCUSSION TITLE
Accessing assets outside lib/ directory

### DISCUSSION THREAD

**Discussion by migalv** [serverpod/serverpod#2986] [Q&A] [answered]
I have some Prompt Templates for LLM generative AI. I decided to save these files in the root directory of the `serverpod_server` package `serverpod_server/promtps/my_prompt_file.txt` 

This was working in development mode, but I've just recently deployed the app to Google Cloud but it's able to find these prompt template files.

Is there any way of accessing files outside the `lib` directory? Is this the proper approach or am I missing something?

**Comment by Isakdl** [comment]
How did you deploy? If you are using the docker container, you have to make sure to copy over any files you would like to keep into the final image.

E.g. add your file/folders like how the copy of the web/migrations folders are done.
```Dockerfile
COPY --from=build /app/bin/server server
COPY --from=build /app/web/ web/
COPY --from=build /app/migrations/ migrations/
```

**Reply by migalv** [reply]
Thanks @Isakdl, let me try

**Reply by migalv** [reply]
Okay this "half worked" because when running the Docker image on a different machine the relative paths based on the current directory don't work anymore.

To me it feels like Serverpod is missing an asset manager similar to Flutter, where you can define assets and configure the base path for the assets based on the environment.

**Comment by migalv** [comment, marked as answer]
I decided to create a `AssetAccessor` that helps when accessing these local assets from different environments.

```dart
import 'dart:io';

import 'package:serverpod/serverpod.dart';
import 'package:path/path.dart' as path;

class ServerpodLocalAssetAccessor {
  ServerpodLocalAssetAccessor._(this._configuredAssetPath);

  static ServerpodLocalAssetAccessor? _instance;

  static ServerpodLocalAssetAccessor get instance {
    if (_instance == null) {
      throw StateError(
        'ServerpodLocalAssetAccessor not initialized. Call initialize() first.',
      );
    }

    return _instance!;
  }

  final String _configuredAssetPath;

  static const String _prodAssetPath = '/assets';
  static const String _stageAssetPath = '/assets';
  static const String _devAssetPath = 'assets';

  // Initialize the handler with Serverpod's runtime config
  static Future<ServerpodLocalAssetAccessor> initialize(String runMode) async {
    if (_instance != null) {
      throw Exception('You cannot initialize the ServerpodLocalAssetAccessor more than once');
    }

    // Get asset path from config based on run mode
    String assetPath;
    switch (runMode) {
      case ServerpodRunMode.production:
        assetPath = _prodAssetPath;
        break;
      case ServerpodRunMode.staging:
        assetPath = _stageAssetPath;
        break;
      case ServerpodRunMode.development:
        assetPath = _devAssetPath;
        break;
      default:
        throw Exception('Unsupported run mode: $runMode');
    }

    _instance = ServerpodLocalAssetAccessor._(assetPath);

    // Ensure the asset directory exists
    await instance.ensureAssetDirectoryExists();

    return instance;
  }

  /// Get the base path for assets
  String get basePath => _configuredAssetPath;

  /// Get the full path for a specific asset
  String getAssetPath(String assetPath) {
    return path.join(basePath, assetPath);
  }

  /// Check if an asset exists
  Future<bool> assetExists(String assetName) async {
    final fullPath = getAssetPath(assetName);
    return await File(fullPath).exists();
  }

  /// Get asset as file
  File getAssetFile(String assetName) {
    return File(getAssetPath(assetName));
  }

  /// Helper method to ensure asset directory exists
  Future<void> ensureAssetDirectoryExists() async {
    final directory = Directory(basePath);
    if (!await directory.exists()) {
      await directory.create(recursive: true);
    }
  }
}
```

Now I can place all my local assets are in the `$PROJECT_HOME/assets/` directory.

Then I updated my Dockerfile with what @Isakdl mentioned.

added: `COPY --from=build /app/assets/ assets/`

If I configure the `ServerpodLocalAssetAccessor` static variables for each environment properly. Then it does not matter where I run the server from I can always access my assets.

To access my assets I just have to `ServerpodLocalAssetAccessor.instance.getAssetFile(${relative_asset_path})`

### For example
I want to access a file in `assets/promtps/summarize_promtp.txt`. Then I can just use the following code: `ServerpodLocalAssetAccessor.instance.getAssetFile('promtps/summarize_promtp.txt')` to get the file.

PS: I tried to use the `config/` files for configuring the assets base path, but it seems they cannot be customized to the project needs yet, so I just added them into the `ServerpodLocalAssetAccessor`

This can be greatly improved, it's just my first take on this. But I believe this should be somehow added to the Serverpod basic features as it's quite easy to add.
